tools/wptserve/wptserve/response.py

```diff
@@ -234,27 +235,46 @@ def write(self):
         self.write_status_headers()
         self.write_content()
 
-    def set_error(self, code, message=""):
+    def set_error(self, code, err=None):
         """Set the response status headers and return a JSON error object:
 
         {"error": {"code": code, "message": message}}
         code is an int (HTTP status code), and message is a text string.
         """
-        err = {"code": code,
-               "message": message}
-        data = json.dumps({"error": err})
+        if 500 <= code < 600:
+            message = self._format_server_error(err)
+            self.logger.warning(message)
+        else:
+            if err is None:
+                message = ""
+            else:
+                message = str(err)
+
+        data = json.dumps({"error": {
+            "code": code,
+            "message": message}
+        })
         self.status = code
         self.headers = [("Content-Type", "application/json"),
                         ("Content-Length", len(data))]
         self.content = data
-        if code == 500:
-            if isinstance(message, str) and message:
-                first_line = message.splitlines()[0]
-            else:
-                first_line = "<no message given>"
-            self.logger.error("Exception loading %s: %s" % (self.request.url,
-                                                            first_line))
-            self.logger.info(message)
+
+    def _format_server_error(self, err):
+        if err is None:
+            suffix = "<no traceback>"
```

Comment:
do we not want to suppress this too? this is effectively a traceback, after all

Reply:
I'm not sure it matters. In practice I don't think we ever raise a 500 error without providing an exception object, but a handler could do that. And if we do end up in this case it's not going to be detected as an unhandled exception by our log parser, so that meets the functional requirements.